File: assemblyscript-algorithms/src/algorithms/search/binary-search.ts
export function binarySearch_u64_asm(
  array: StaticArray<u64>,
  target: u64
): i32 {
  let l = 0;
  let r = array.length - 1;
  let mid = -1;
  while (r >= l) {
    mid = l + (r - l) / 2;
    if (array[mid] == target) {
      return mid;
    }
    if (array[mid] > target) {
      r = mid - 1;
    } else {
      l = mid + 1;
    }
  }
  return -1;
}
export function binarySearch_i64_asm(
  array: StaticArray<i64>,
  target: i64
): i32 {
  let l = 0;
  let r = array.length - 1;
  let mid = -1;
  while (r >= l) {
    mid = l + (r - l) / 2;
    if (array[mid] == target) {
      return mid;
    }
    if (array[mid] > target) {
      r = mid - 1;
    } else {
      l = mid + 1;
    }
  }
  return -1;
}
export function binarySearch_f64_asm(
  array: StaticArray<f64>,
  target: f64
): i32 {
  let l = 0;
  let r = array.length - 1;
  let mid = -1;
  while (r >= l) {
    mid = l + floor((r - l) / 2);
    if (array[mid] == target) {
      return mid;
    }
    if (array[mid] > target) {
      r = mid - 1;
    } else {
      l = mid + 1;
    }
  }
  return -1;
}
export function binarySearch_f32_asm(
  array: StaticArray<f32>,
  target: f32
): i32 {
  let l = 0;
  let r = array.length - 1;
  let mid = -1;
  while (r >= l) {
    mid = l + floor((r - l) / 2);
    if (array[mid] == target) {
      return mid;
    }
    if (array[mid] > target) {
      r = mid - 1;
    } else {
      l = mid + 1;
    }
  }
  return -1;
}
export function binarySearch_u32_asm(
  array: StaticArray<u32>,
  target: u32
): i32 {
  let l = 0;
  let r = array.length - 1;
  let mid = -1;
  while (r >= l) {
    mid = l + (r - l) / 2;
    if (array[mid] == target) {
      return mid;
    }
    if (array[mid] > target) {
      r = mid - 1;
    } else {
      l = mid + 1;
    }
  }
  return -1;
}
export function binarySearch_i32_asm(
  array: StaticArray<i32>,
  target: i32
): i32 {
  let l = 0;
  let r = array.length - 1;
  let mid = -1;
  while (r >= l) {
    mid = l + (r - l) / 2;
    if (array[mid] == target) {
      return mid;
    }
    if (array[mid] > target) {
      r = mid - 1;
    } else {
      l = mid + 1;
    }
  }
  return -1;
}
export function binarySearch_u16_asm(
  array: StaticArray<u16>,
  target: u16
): i32 {
  let l = 0;
  let r = array.length - 1;
  let mid = -1;
  while (r >= l) {
    mid = l + (r - l) / 2;
    if (array[mid] == target) {
      return mid;
    }
    if (array[mid] > target) {
      r = mid - 1;
    } else {
      l = mid + 1;
    }
  }
  return -1;
}
export function binarySearch_i16_asm(
  array: StaticArray<i16>,
  target: i16
): i32 {
  let l = 0;
  let r = array.length - 1;
  let mid = -1;
  while (r >= l) {
    mid = l + (r - l) / 2;
    if (array[mid] == target) {
      return mid;
    }
    if (array[mid] > target) {
      r = mid - 1;
    } else {
      l = mid + 1;
    }
  }
  return -1;
}
export function binarySearch_u8_asm(
  array: StaticArray<i8>,
  target: i8
): i32 {
  let l = 0;
  let r = array.length - 1;
  let mid = -1;
  while (r >= l) {
    mid = l + (r - l) / 2;
    if (array[mid] == target) {
      return mid;
    }
    if (array[mid] > target) {
      r = mid - 1;
    } else {
      l = mid + 1;
    }
  }
  return -1;
}
export function binarySearch_i8_asm(array: StaticArray<u8>, target: u8): i32 {
  let l = 0;
  let r = array.length - 1;
  let mid = -1;
  while (r >= l) {
    mid = l + (r - l) / 2;
    if (array[mid] == target) {
      return mid;
    }
    if (array[mid] > target) {
      r = mid - 1;
    } else {
      l = mid + 1;
    }
  }
  return -1;
}

File: c-algorithms/src/search/binary-search/binary-search.c
#include <emscripten.h>
#include "search/binary-search/binary-search.h";
#include "shared/shared.h"

EMSCRIPTEN_KEEPALIVE
int binarySearch_u64_c(Array1d_u64* array, u64 target) {
    int l = 0;
    int r = array->length - 1;
    int mid = -1;
    while(r >= l) {
        mid = l + (r - l) / 2;
        if (array->items[mid] == target) {
            return mid;
        }
        if (array->items[mid] > target) {
            r = mid - 1;
        }
        else {
            l = mid + 1;
        }
    }
    return -1;
}
EMSCRIPTEN_KEEPALIVE
int binarySearch_i64_c(Array1d_i64* array, i64 target) {
    int l = 0;
    int r = array->length - 1;
    int mid = -1;
    while(r >= l) {
        mid = l + (r - l) / 2;
        if (array->items[mid] == target) {
            return mid;
        }
        if (array->items[mid] > target) {
            r = mid - 1;
        }
        else {
            l = mid + 1;
        }
    }
    return -1;
}
EMSCRIPTEN_KEEPALIVE
int binarySearch_f64_c(Array1d_f64* array, f64 target) {
    int l = 0;
    int r = array->length - 1;
    int mid = -1;
    while(r >= l) {
        mid = l + (r - l) / 2;
        if (array->items[mid] == target) {
            return mid;
        }
        if (array->items[mid] > target) {
            r = mid - 1;
        }
        else {
            l = mid + 1;
        }
    }
    return -1;
}
EMSCRIPTEN_KEEPALIVE
int binarySearch_f32_c(Array1d_f32* array, f32 target) {
    int l = 0;
    int r = array->length - 1;
    int mid = -1;
    while(r >= l) {
        mid = l + (r - l) / 2;
        if (array->items[mid] == target) {
            return mid;
        }
        if (array->items[mid] > target) {
            r = mid - 1;
        }
        else {
            l = mid + 1;
        }
    }
    return -1;
}
EMSCRIPTEN_KEEPALIVE
int binarySearch_u32_c(Array1d_u32* array, u32 target) {
    int l = 0;
    int r = array->length - 1;
    int mid = -1;
    while(r >= l) {
        mid = l + (r - l) / 2;
        if (array->items[mid] == target) {
            return mid;
        }
        if (array->items[mid] > target) {
            r = mid - 1;
        }
        else {
            l = mid + 1;
        }
    }
    return -1;
}
EMSCRIPTEN_KEEPALIVE
int binarySearch_i32_c(Array1d_i32* array, i32 target) {
    int l = 0;
    int r = array->length - 1;
    int mid = -1;
    while(r >= l) {
        mid = l + (r - l) / 2;
        if (array->items[mid] == target) {
            return mid;
        }
        if (array->items[mid] > target) {
            r = mid - 1;
        }
        else {
            l = mid + 1;
        }
    }
    return -1;
}
EMSCRIPTEN_KEEPALIVE
int binarySearch_u16_c(Array1d_u16* array, u16 target) {
    int l = 0;
    int r = array->length - 1;
    int mid = -1;
    while(r >= l) {
        mid = l + (r - l) / 2;
        if (array->items[mid] == target) {
            return mid;
        }
        if (array->items[mid] > target) {
            r = mid - 1;
        }
        else {
            l = mid + 1;
        }
    }
    return -1;
}
EMSCRIPTEN_KEEPALIVE
int binarySearch_i16_c(Array1d_i16* array, i16 target) {
    int l = 0;
    int r = array->length - 1;
    int mid = -1;
    while(r >= l) {
        mid = l + (r - l) / 2;
        if (array->items[mid] == target) {
            return mid;
        }
        if (array->items[mid] > target) {
            r = mid - 1;
        }
        else {
            l = mid + 1;
        }
    }
    return -1;
}
EMSCRIPTEN_KEEPALIVE
int binarySearch_u8_c(Array1d_u8* array, u8 target) {
    int l = 0;
    int r = array->length - 1;
    int mid = -1;
    while(r >= l) {
        mid = l + (r - l) / 2;
        if (array->items[mid] == target) {
            return mid;
        }
        if (array->items[mid] > target) {
            r = mid - 1;
        }
        else {
            l = mid + 1;
        }
    }
    return -1;
}
EMSCRIPTEN_KEEPALIVE
int binarySearch_i8_c(Array1d_i8* array, i8 target) {
    int l = 0;
    int r = array->length - 1;
    int mid = -1;
    while(r >= l) {
        mid = l + (r - l) / 2;
        if (array->items[mid] == target) {
            return mid;
        }
        if (array->items[mid] > target) {
            r = mid - 1;
        }
        else {
            l = mid + 1;
        }
    }
    return -1;
}
